drivers: can: Rework can_configure API and zcan_frame

alexanderwachter opened the pull request:
The previous API can't change the sampling-point and only allowed
bitrates that fit the time segments.
The new API allows for shifting the sampling-point and adjusts the
number of time quantum in a bit to all more possible bitrates.
The functions to calculate the timings are moved to the can_common file.
They can be used for all drivers.

alexanderwachter committed:
drivers: can: Rework can_configure API

The previous API can't change the sampling-point and only allowed
bitrates that fit the time segments.
The new API allows for shifting the sampling-point and adjusts the
number of time quantum in a bit to all more possible bitrates.
The functions to calculate the timings are moved to the can_common file.
They can be used for all drivers.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter commented on drivers/can/can_mcp2515.c:
```diff
@@ -290,42 +290,45 @@ static int mcp2515_get_mode(const struct device *dev, uint8_t *mode)
 	/* CNF1; SJW<7:6> | BRP<5:0> */
-	uint8_t brp = (dev_cfg->osc_freq / (bit_length * bitrate * 2)) - 1;
-	const uint8_t sjw = (dev_cfg->tq_sjw - 1) << 6;
+	uint8_t brp = timing->prescaler;
```
Whoops, need to check for NULL.

karstenkoenig replied:
Just in case, this applies to all existing drivers :-)

alexanderwachter commented:
This PR is the first step in the direction of a CAN-FD API.

martinjaeger commented on include/drivers/can.h:
```diff
@@ -508,28 +560,208 @@ static inline void z_impl_can_detach(const struct device *dev, int filter_id)
+ * @retval -EINVAL bitrate cannot be reached.
+ * @retval -EIO General input / output error, failed to configure device.
+ */
+static inline int can_set_bitrate(struct device *dev, uint32_t bitrate)
```
```suggestion
static inline int can_set_bitrate(const struct device *dev, uint32_t bitrate)
```
(same for the functions below)

martinjaeger commented on dts/bindings/can/can-controller.yaml:
```diff
@@ -23,13 +23,20 @@ properties:
+      description: >
+        Sample point in permille.
+        This param is required if segments are not given.
+        If the sample point is give, the segments are ignored.
```
```suggestion
        If the sample point is given, the segments are ignored.
```

martinjaeger reviewed (commented):
Only got some nits comments so far.

karstenkoenig commented on drivers/can/can_common.c:
```diff
@@ -11,6 +11,9 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(can_driver);
 
+#define CAN_CLAMP(val, min, max) MIN(MAX(val, min), max)
```
This looks fairly generic, maybe that'd be nice to have in `util.h`?

martinjaeger replied:
Just seeing PR https://github.com/zephyrproject-rtos/zephyr/pull/29576, which should be used once merged.

alexanderwachter replied:
Sure. Didn't knew that it is already in! Thanks for the hint.

martinjaeger commented:
As agreed with @alexanderwachter on Slack, I'm taking over this PR and will fix remaining issues and adapt it to updates happened in Zephyr upstream.

Instead of changing the `can_configure` function, I would like to suggest a different approach.

### Approach for update of API

1. Add two new functions to `can_driver_api` with the following public API: 

```C
int can_set_mode(const struct device *dev, enum can_mode mode);

int can_set_timing(const struct device *dev, const struct can_timing *timing, 
                   const struct can_timing *timing_data);
```
These two functions would be implemented in each driver instead of `can_configure`.

The function `can_set_timing` would be the same for classic CAN and CAN-FD, but in case of classic CAN, the parameter `timing_data` is set to `NULL` and will be ignored.

2. Add new function to `can.h`:

```C
static inline can_set_bitrate(const struct device *dev, uint32_t bitrate, uint32_t bitrate_data) { ... }
```
which calls `can_set_timing` with calculated parameters. Once again, `bitrate_data` is only relevant for CAN-FD and is ignored for classic CAN.

3. Implement `can_configure(...)` as `static inline` function in `can.h` (instead of implementation in the driver) which calls above `can_set_mode` and `can_set_bitrate` functions.

### Advantages

- Public API `can_configure` is kept as it is. We could think about deprecating it, as it is only suitable for classic CAN.
- `CAN_NO_MODE_CHANGE` is not needed.
- More straightforward implementation in the driver: Instead of selecting between different cases in `can_configure` indicated by passing `NULL` or above `CAN_NO_MODE_CHANGE` we have a dedicated function for mode change and timing settings.

I would like to know if you are generally fine with this idea. If yes, I'll go ahead and propose the code for further discussion.

alexanderwachter commented:
@martinjaeger you are right! Using two functions make more sense!

karstenkoenig commented:
Yes agreed, dedicated functions with less checks for null/null equivalent sounds better

henrikbrixandersen commented:
> I would like to know if you are generally fine with this idea. If yes, I'll go ahead and propose the code for further discussion.

@martinjaeger This sounds like a sane approach. I'd love to review the suggested code changes.

martinjaeger commented:
Great that you all agree with the proposed change. Yesterday, @alexanderwachter and I made made quite good progress on this PR already. We will push the updates soon.

alexanderwachter commented:
@martinjaeger updated the PR with the changes from your branch.
@henrikbrixandersen @karstenkoenig can you review? I'll squash the commits once the PR is ready.
Does someone have an idea of how we can handle the timings from the device tree better?
Should we drop manual timing settings in favor of bitrate and sampling-point?

alexanderwachter commented on include/drivers/can.h:
```diff
@@ -508,28 +560,208 @@ static inline void z_impl_can_detach(const struct device *dev, int filter_id)
+		return -EINVAL;
+	}
+
+	ret = can_calc_timing_data(dev, &timing_data, bitrate, 875);
```
Yep

martinjaeger commented on drivers/can/can_handlers.c:
```diff
@@ -7,16 +7,35 @@
-				    (uint32_t)bitrate);
+	return z_impl_can_get_core_clock((struct device *)dev);
 }
 #include <syscalls/can_configure_mrsh.c>
```
```suggestion
#include <syscalls/can_get_core_clock_mrsh.c>
```

martinjaeger commented on drivers/can/can_mcux_flexcan.c:
```diff
@@ -91,36 +92,73 @@ struct mcux_flexcan_data {
+static int mcux_flexcan_set_timing(const struct device *dev,
+				   const struct can_timing *timing,
+				   const struct can_timing *timing_data)
+{
```
```suggestion
{
	ARG_UNUSED(timing_data);
```

martinjaeger commented on drivers/can/can_mcux_flexcan.c:
```diff
@@ -91,36 +92,73 @@ struct mcux_flexcan_data {
+{
+	struct mcux_flexcan_data *data = dev->data;
+	const struct mcux_flexcan_config *config = dev->config;
+	flexcan_config_t flexcan_config;
```
```suggestion
```
I think you can update `data->timing` instead and also pass it to `FLEXCAN_Init` at the end. Requires less updates of member variables below, as they are already in the correct state. And it preserves the mode if timing is changed afterwards.

alexanderwachter replied:
Not sure what you meant here

martinjaeger replied:
Oops, never mind. I confused `flexcan_config_t` and `flexcan_timing_config_t` and thought that the timing parameter of the `mcux_flexcan_data` could be re-used here. But looking twice it doesn't make sense.

martinjaeger commented on drivers/can/can_mcux_flexcan.c:
```diff
@@ -633,7 +672,37 @@ static int mcux_flexcan_init(const struct device *dev)
+		return -EINVAL;
+	}
+
+	data->timing.preDivider = timing.prescaler;
+	data->timing.rJumpwidth = timing.sjw;
+	data->timing.phaseSeg1 = timing.phase_seg1;
+	data->timing.phaseSeg2 = timing.phase_seg2;
+	data->timing.propSeg = timing.prop_seg;
```
Maybe call `mcux_flexcan_set_timing(dev, timing)` instead?

alexanderwachter replied:
That would call the set bitrate on an uninitialized controller. I'd avoid that

martinjaeger replied:
Ok, makes sense.

martinjaeger commented on include/drivers/can.h:
```diff
@@ -98,7 +98,7 @@ enum can_mode {
 	CAN_LOOPBACK_MODE,
 	/*Combination of loopback and silent*/
-	CAN_SILENT_LOOPBACK_MODE
+	CAN_SILENT_LOOPBACK_MODE,
```
Nits:
```suggestion
	CAN_SILENT_LOOPBACK_MODE
```

martinjaeger reviewed (changes requested):
Thanks @alexanderwachter for integrating the updates from my branch and feel free to squash them in to clean up the commit history.

In theory, the changes in the CAN shell, the tests, ISO-TP and CANopen are not required anymore, as the `can_configure` function is kept as it is. So I suggest to put them into a separate commit.

alexanderwachter committed:
drivers: can: rework zcan_frame and zcan_filter

Reordering of the struct elements to match the Linux format.
The __packed() is not necessary anymore.
std_id and ext_id is merged to id in the frame and filter.
Additionally, the frames are ready for CAN-FD.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
tests: canbus: remove unnecessary test for invalid frames

This test is now obsolete, because there is no std and ext
id anymore.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter commented:
@henrikbrixandersen @martinjaeger @karstenkoenig 
 I'm not sure if commit `drivers: can: extend can_frame to allow external data buffers` is a good idea.
This needs to be discussed. It would allow us to have bit compatible frames in the future that only hold the pointer.
For classic CAN, it is not worth it, but for CAN-FD with 64 bytes data, it could make sense.

martinjaeger commented on include/drivers/can.h:
```diff
@@ -167,34 +167,49 @@ struct can_filter {
+	/** @endcond */
 #endif
-} __packed;
+	/** The frame payload data. */
+	union {
+		/** frame data, if ext_buf is unset */
+		uint8_t data[CAN_MAX_DLEN];
+		uint32_t data_32[ceiling_fraction(CAN_MAX_DLEN, sizeof(uint32_t))];
+		/** pointer ot the frame data, if ext_buf is set */
+		const uint8_t *buf;
+	};
```
Should `CAN_MAX_DLEN` become Kconfig-urable for CAN-FD?

As mentioned in the issue, I'm not a huge fan of unions (MISRA even forbids them, btw.). With up to 64 bytes in CAN-FD we would need `memcpy` instead of direct `uint32` copying anyway. So I would probably just use the most simple (and endian-safe) approach: `uint8_t data[CONFIG_CAN_FRAME_MAX_DLEN];`

I do see the benefit of using a pointer instead of the buffer to avoid copy operations in longer streams of data (e.g. for CAN-FD or ISO-TP). However, if each driver has to support both buffer and pointer frame types, this can create quite a lot of noise in the code. If we sacrifice 4 more bytes, we could do something like this (if we ensure that `data` points to the right location during initialization):

```C
#ifdef CONFIG_CAN_FRAME_INT_BUF
	uint8_t data_buf[CONFIG_CAN_FRAME_MAX_DLEN];
#endif
	const uint8_t *data;
```
The drivers could just always use the `data` pointer. Not totally sure if that makes sense...

alexanderwachter replied:
I would not add anything after the data, because we would lose the ability to cast shorter frames to z_can frames.
data_buf should always be present because receiving frames will always put the data in there.
We could still drop that commit.
Regarding the uint32_t data_32, I'd like to keep them, because otherwise, we would need to use casts in the drivers, which I generally don't like.

alexanderwachter replied:
Yes, CAN_MAX_DLEN is for CAN-FD

martinjaeger replied:
I'm fine with the union to allow all different options to store the message. But don't we have to rewrite all drivers to support ext_buf? So this might be something for a separate PR.

alexanderwachter replied:
Removed that commit.

alexanderwachter replied:
Removed.

alexanderwachter committed:
drivers: can: Compile time check of can timing not set

Check if the timing is not set from device tree and exclude
sample-point algorithm if it is not used from device-tree.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

martinjaeger commented on include/canbus/isotp.h:
```diff
@@ -103,7 +103,7 @@ struct isotp_msg_id {
 	/** Message identifier*/
 	union {
 		uint32_t std_id  : 11;
-		uint32_t ext_id  : 29;
+		uint32_t id  : 29;
 	};
```
Any special reason to keep `std_id` here?

alexanderwachter replied:
Here we can keep the union for convenience. As long as we don't use another bit of the uint32_t, we can use the union.

alexanderwachter replied:
Renamed back to ext_id. This was a mistake.

martinjaeger commented on include/drivers/can.h:
```diff
@@ -167,34 +167,45 @@ struct can_filter {
-} __packed;
+	/** The frame payload data. */
+	union {
+		/** frame data, if ext_buf is unset */
```
This comment is not valid anymore, as `ext_buf` is not existing anymore.

alexanderwachter replied:
Removed comment.

martinjaeger reviewed (commented):
The commit message of `drivers: can: rework zcan_frame and zcan_filter` still states "... and the data field can hold either the data or a pointer to the data.", which is not the case anymore.

Cherish-forever commented:
I have been verify this patch. 2 CAN, each run 1Mbit/s with protocol stack.
Thank you very much
@alexanderwachter

alexanderwachter commented:
@Cherish-forever Thanks for testing!

martinjaeger reviewed (approved):
Thanks a lot @alexanderwachter. Looks good to me and brings us a big step closer to get the CAN-FD PR in.

Would be great if @henrikbrixandersen and @karstenkoenig could also have another look at this.

karstenkoenig commented on dts/bindings/can/can-fd-controller.yaml:
```diff
@@ -0,0 +1,32 @@
+      type: int
+      required: false
+      description: >
+        Sample point in permille.
```
'for the data phase' could be added here to be in line with the time quantums

alexanderwachter replied:
Done.

karstenkoenig commented on dts/bindings/can/can-fd-controller.yaml:
```diff
@@ -0,0 +1,32 @@
+      description: >
+        Sample point in permille.
+        This param is required if segments are not given.
+        If the sample point is give, the segments are ignored.
```
small typo, given

alexanderwachter replied:
Done.

karstenkoenig reviewed (commented):
Looks great! Did you do a sanity check on the mcp2515 by chance? If not I can try and find my devkit in the moving crates today and also give that a spin :-)

alexanderwachter commented on drivers/can/can_common.c:
```diff
@@ -143,3 +142,120 @@ int can_attach_workq(const struct device *dev, struct k_work_q *work_q,
+			 uint32_t bitrate, uint16_t sample_pnt)
+{
+	const uint32_t core_clock = can_get_core_clock(dev);
+	const struct can_driver_api *api = dev->driver_api;
```
Thx. Missed that. That would hit CI when we merge the first CAN-FD driver :-)

alexanderwachter commented on drivers/can/can_mcp2515.c:
```diff
@@ -17,6 +17,23 @@ LOG_MODULE_REGISTER(mcp2515_can);
 
 #include "can_mcp2515.h"
 
+#define SP_IS_SET(inst) DT_INST_NODE_HAS_PROP(inst, sample_point) ||
```
Unfortunately not. I'm using DT_INST_ macro, where the DT_DRV_COMPAT macro.

karstenkoenig replied:
Aight, thanks for clearing it up. Having a dejavu here, did I ask this before? :-D

alexanderwachter commented on drivers/can/can_mcp2515.c:
```diff
@@ -550,11 +573,11 @@ static uint8_t mcp2515_filter_match(struct zcan_frame *msg,
 
 	if (msg->id_type == CAN_STANDARD_IDENTIFIER) {
-		if ((msg->std_id ^ filter->std_id) & filter->std_id_mask) {
+		if ((msg->id ^ filter->id) & filter->id_mask) {
```
Removed.

alexanderwachter commented:
@karstenkoenig did not run it so far. Please do it if you find the time.
Thanks a lot for the review!

karstenkoenig commented:
the tests fail on mcp2515 :-(

> W: Bitrate error: 8
> ASSERTION FAIL [(timing->prop_seg >= 1) && (timing->prop_seg <= 8)] @ WEST_TOPDIR/zephyr/drivers/can/can_mcp2515.c:366
>        1 <= PROP <= 8

I am sorry I am out for a long weekend trip and still have to pack but I promise to take a look into it early next week.

alexanderwachter commented:
@karstenkoenig can you run the test again. It was a copy and paste error. Sry!
`timing.prop_seg = dev_cfg->tq_prop;` was `timing.prop_seg = 0;` (copy from stm32)

karstenkoenig commented:
> @karstenkoenig can you run the test again. It was a copy and paste error. Sry!
> `timing.prop_seg = dev_cfg->tq_prop;` was `timing.prop_seg = 0;` (copy from stm32)

Yes, it works now! Thanks

alexanderwachter commented:
@karstenkoenig did you check with the new algorithm? You need to set a sample-point in the device tree to activate it.

henrikbrixandersen reviewed (changes requested):
As mentioned earlier, I still believe this PR bundles too many changes in one (CAN-FD preparations, sampling point changes, API changes, frame struct changes, ...). What is the reasoning behind not splitting these up in smaller chunks?

alexanderwachter commented:
Because I want to bundle the API break in one PR. Still, the commits do only one thing.

alexanderwachter commented:
The config API change comes from the work on CAN-FD. That is why it already supports CAN-FD

karstenkoenig commented:
> @karstenkoenig did you check with the new algorithm? You need to set a sample-point in the device tree to activate it.

I had not, but now did. It works for sample points above about 55%, which seems sensible. Below it hits an assert for `timing->prop_seg + timing->phase_seg1 >= timing->phase_seg2`. Would other CAN controllers support sampling rates below 55%? Maybe could just range check the input of `sample-point`, it also allows values over 1000.

alexanderwachter commented:
A check for > 1000 males sense!
The assert you get for sample points below 55% depends on the baudrate and the crystal. The algorithm needs to adjust the prescaler and (prop dalay + ts1 + ts2 + 1) so that it matches the baudrate exactly. That limits the samplepoint somehow.

alexanderwachter committed:
tests: drivers: can: timing

Added tests for the timing algorithm.
The tests calculate timings for some bitrates with sample-points
and verify the results.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter commented:
@karstenkoenig added the checks.
I also added a test case for the sample-point calculation.
This test can either run on HW or native_psix.

pabigot commented on include/drivers/can.h:
```diff
@@ -98,7 +98,7 @@ enum can_mode {
 	/*Controller is in loopback mode (receive own messages)*/
 	CAN_LOOPBACK_MODE,
 	/*Combination of loopback and silent*/
-	CAN_SILENT_LOOPBACK_MODE,
```
It's generally nice to leave the comma on the last element, as it removes gratuitous diffs if new elements are added below.

pabigot commented on include/drivers/can.h:
```diff
@@ -167,34 +167,44 @@ struct can_filter {
+	uint8_t res1;  /* reserved / padding */
+	/** @endcond */
 #endif
-} __packed;
```
Are you sure it's safe to remove `__packed`?  Two concerns:

* Without this the placement of an instance of this type within a buffer must be naturally aligned in order for accesses to work.  Can you be sure that all instances will be aligned?
* As was pointed out to me in another context, the C standard specifies that padding can occur between members and at the end.  While a quality implementation should not introduce unnecessary padding, the standard doesn't seem to prohibit it.

I think the first point is a concern; the second might be too picky at least for common architectures.

alexanderwachter replied:
this frame is not sent over the wire.

pabigot commented on include/drivers/can.h:
```diff
@@ -167,34 +167,44 @@ struct can_filter {
-	 */
-	uint32_t id_type : 1;
+	/** Message identifier*/
+	uint32_t id      : 29;
```
Technically bitfield types can only be `int`, `unsigned int`, and `bool`.  Zephyr doesn't pay much attention to this.

pabigot commented on include/drivers/can.h:
```diff
@@ -167,34 +167,44 @@ struct can_filter {
 	 * The timer is incremented every bit time and captured at the start
 	 * of frame bit (SOF).
 	 */
 	uint16_t timestamp;
```
Isn't this a 2-byte value placed 3 bytes after a 4 byte value, which violates the natural alignment requirements?

alexanderwachter replied:
It's two bytes after two one-byte values.

pabigot reviewed (commented):
This is outside my area, just a few comments about packing.

henrikbrixandersen commented on drivers/can/can_loopback.c:
```diff
@@ -234,19 +229,39 @@ static void can_loopback_register_state_change_isr(const struct device *dev,
 
+uint32_t can_loopback_get_core_clock(const struct device *dev)
+{
+	return 16000000;
```
Perhaps a comment on why this seemingly arbitrary clock rate was chosen would be nice for future reference.

alexanderwachter replied:
It's an arbitrary realistic value for the tescases. I left a comment.

henrikbrixandersen commented on include/drivers/can.h:
```diff
@@ -293,6 +340,8 @@ typedef enum can_state (*can_get_state_t)(const struct device *dev,
 typedef void(*can_register_state_change_isr_t)(const struct device *dev,
 					       can_state_change_isr_t isr);
 
+typedef uint32_t (*can_get_core_clock_t)(const struct device *dev);
```
```suggestion
typedef int (*can_get_core_clock_t)(const struct device *dev, uint32_t *rate);
```

Being able to return an error is important. It may be that communication with an external CAN controller for reading the clock rate failed, but with the proposed function signature, we have no way of propagating this (indeed, some of the drivers in this PR simply ignore this an return `0` in case of failure).

alexanderwachter replied:
True. It makes sense to use the same signature as clock_control_get_rate.

henrikbrixandersen commented on include/drivers/can.h:
```diff
@@ -509,28 +569,190 @@ static inline void z_impl_can_detach(const struct device *dev, int filter_id)
+
+#ifdef CONFIG_CAN_FD_MODE
+/**
+ * @brief Calculate timing parameters for the data pase
```
```suggestion
 * @brief Calculate timing parameters for the data phase
```

henrikbrixandersen commented on include/drivers/can.h:
```diff
@@ -509,28 +569,190 @@ static inline void z_impl_can_detach(const struct device *dev, int filter_id)
+ * @param dev        Pointer to the device structure for the driver instance.
+ * @param res        Result is written into the can_timing struct provided.
+ * @param bitrate    Target bitrate for the datapase in bits/s
+ * @param sample_pnt Sampling point datapase in permill of the entire bit time.
```
```suggestion
 * @param sample_pnt Sampling point data phase in permille of the entire bit time.
```

henrikbrixandersen commented on include/drivers/can.h:
```diff
@@ -509,28 +569,190 @@ static inline void z_impl_can_detach(const struct device *dev, int filter_id)
+ * @param timing  Result is written into the can_timing struct provided.
+ * @param bitrate Target bitrate.
+ *
+ * @retval bitrate error
```
I think this needs a bit of elaboration.

henrikbrixandersen commented on include/drivers/can.h:
```diff
@@ -509,28 +569,190 @@ static inline void z_impl_can_detach(const struct device *dev, int filter_id)
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_CAN_FD_MODE
+	ret = can_calc_timing_data(dev, &timing_data, bitrate_data, 875);
+	if (ret < 0) {
+		return -EINVAL;
+	}
+
+	return can_set_timing(dev, &timing, NULL);
+#else
+	return can_set_timing(dev, &timing, &timing_data);
+#endif /* CONFIG_CAN_FD_MODE */
```
Is this correct? Shouldn't the case where CAN-FD is enabled use both timing parameter structs (and not the other way around)?

alexanderwachter replied:
Yes, that's wrong.

talih0 commented on include/drivers/can.h:
```diff
@@ -509,28 +562,196 @@ static inline void z_impl_can_detach(const struct device *dev, int filter_id)
+	return can_set_timing(dev, &timing, &timing_data);
+#else
+	return can_set_timing(dev, &timing, NULL);
+#endif /* CONFIG_CAN_FD_MODE */
```
This forgets to set timing.sjw and timing_data.sjw before calling can_set_timing()

talih0 replied:
@alexanderwachter If you have time, could you have a look pls? ^^
Should I open an issue?

alexanderwachter replied:
@talih0  It's a bug, yes. Can you open an issue?
